### PR TITLE
Add image decoder registry.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
The purpose of this change is to make it possible to use all of the available image decoders without enumerating them by hand at use site (which is at odds with the goal of abstracting away the particular decoders from their usage).

- `ImageFormatRegistry` is now a singleton.
- `ImageFormatRegistry` has a format lookup by name
- a mutable `ImageFormat` can be obtained from the registry
- There's a registration macro for decoders, which registers a decoder factory in the (now global) `ImageFormatRegistry`.

### Affected modules and functionalities:
- image format
- image format registry
- imgcodec/decoders (to a small extent)

### Tests:

- [ ] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2984

### CODE REVIEW

**Review by jantonguirao — APPROVED** [review]

File: include/dali/imgcodec/image_format.h [hunk 1]
```diff
@@ -71,32 +71,34 @@ class DLL_PUBLIC ImageFormat {
...
-  void RegisterDecoder(std::shared_ptr<ImageDecoder> decoder, float priority);
+  void RegisterDecoder(std::shared_ptr<ImageDecoderFactory> factory, float priority);
 
  private:
   std::string name_;
   std::shared_ptr<ImageParser> parser_;
-  std::multimap<float, std::shared_ptr<ImageDecoder>> decoders_;
-  std::vector<ImageDecoder*> decoder_ptrs_;
+  std::multimap<float, std::shared_ptr<ImageDecoderFactory>> decoders_;
+  std::vector<ImageDecoderFactory*> decoder_ptrs_;
```
Comment (prak-nv) [inline include/dali/imgcodec/image_format.h:101]: why do you need it? You can iterate over map.

File: include/dali/imgcodec/image_format.h [hunk 2]
```diff
@@ -71,32 +71,34 @@ class DLL_PUBLIC ImageFormat {
...
+   * @param factory   A decoder factory
    * @param priority  A float representing the priority of this codec.
    *                  The lower the number, the higher priority the codec has.
    */
-  void RegisterDecoder(std::shared_ptr<ImageDecoder> decoder, float priority);
+  void RegisterDecoder(std::shared_ptr<ImageDecoderFactory> factory, float priority);
 
  private:
   std::string name_;
   std::shared_ptr<ImageParser> parser_;
```
Comment (prak-nv) [inline include/dali/imgcodec/image_format.h:99]: Dont get why it is shared_ptr

Comment (mzient) [reply include/dali/imgcodec/image_format.h:99]: Yeah, in principle, it doesn't need to be - but what would be the alternative? `unique_ptr` + `bool own`?

File: include/dali/imgcodec/image_format.h [hunk 3]
```diff
@@ -71,32 +71,34 @@ class DLL_PUBLIC ImageFormat {
...
    * @brief Returns a string representing the name of the format
    */
   const std::string& Name() const;
 
   /**
-   * @brief Returns a set of decoders associated with this format
+   * @brief Returns a set of decoder factories associated with this format
    */
-  span<ImageDecoder* const> Decoders() const;
+  span<ImageDecoderFactory *const> Decoders() const;
```
[the remark this answers is not in the document]
Comment (mzient) [reply include/dali/imgcodec/image_format.h:84]: True, but it's not something new in this PR. I can think about it later.

**Pull request merged**